r/src/array.cpp

```diff
@@ -210,6 +210,20 @@ std::shared_ptr<Array> MakeStringArray(StringVector_ vec) {
   return MakeArray(data);
 }
 
+std::shared_ptr<Array> MakeFactorArray(Rcpp::IntegerVector_ factor) {
+  auto dict_values = MakeStringArray(Rf_getAttrib(factor, R_LevelsSymbol));
+  auto dict_type = dictionary(int32(), dict_values, Rf_inherits(factor, "ordered"));
+
+  auto n = Rf_length(factor);
+  IntegerVector r_indices(n);
+  for (R_xlen_t i = 0; i < n; i++) {
+    r_indices[i] = factor[i] == NA_INTEGER ? NA_INTEGER : (factor[i] - 1);
```

Comment:
I would think that you would want to allocate an output array and write directly into that rather than going through an intermediate step. Then the NA values don't have to be double-checked

Reply:
Right. I’ll update so that we only check NAs once. 

I’ll have to duplicate some of the code of SimpleArray, but that’s fine.

Reply:
✅